doc/ractor.md

```diff
@@ -0,0 +1,882 @@
+
+### Closing Ractor's ports
+
+* `Ractor#close_incoming/outgoing` close incoming/outgoing ports (similar to `Queue#close`).
+* `Ractor#close_incoming`
+  * `r.send(obj) ` where `r`'s incoming port is closed, will raise an exception.
+  * When the incoming queue is empty and incoming port is closed, `Ractor.recv` raise an exception. If incoming queue is not empty, it dequeues an object.
+* `Ractor#close_outgoing`
+  * `Ractor.yield` on a Ractor which closed the outgoing port, it will raise an exception.
+  * `Ractor#take` for a Ractor which closed the outgoing port, it will raise an exception. If `Ractor#take` is blocking, it will raise an exception.
+* When a Ractor terminates, the ports are closed automatically.
+  * Return value of the Ractor's block will be yield as `Ractor.yield(ret_val)`, even if the implementation terminate the based native thread.
+
+
+Example (try to take from closed Ractor):
+
+```ruby
+  r = Ractor.new do
+    'finish'
+  end
+  r.take # success (will return 'finish')
+  begin
+    o = r.take # try to take from closed Ractor
+  rescue Ractor::ClosedError
+    'ok'
+  else
+    "ng: #{o}"
+  end
+```
+
+Example (try to send to closed (terminated) Ractor):
+
+```ruby
+  r = Ractor.new do
+  end
+
+  r.take # wait terminate
+
+  begin
+    r.send(1)
+  rescue Ractor::ClosedError
+    'ok'
+  else
+    'ng'
+  end
+```
+
+When multiple Ractors waiting for `Ractor.yield()`, `Ractor#close_outgoing` will cancel all blocking by raise an exception (`ClosedError`).
+
+### Send a message by copying
+
+`Ractor#send(obj)` or `Ractor.yield(obj)` copy `obj` deeply if `obj` is an unshareable object.
+
+```ruby
+obj = 'str'.dup
+r = Ractor.new obj do |msg|
+  # return received msg's object_id
+  msg.object_id
+end
+  
+obj.object_id == r.recv #=> false
+```
+
+Current implementation uses Marshal protocol (similar to dRuby). We can not send Marshal unsupported objects.
+
+```ruby
+obj = Thread.new{}
+begin
+  r = Ractor.new obj do |msg|
+    msg
+  end
+rescue TypeError => e
+  e.message #=> no _dump_data is defined for class Thread
+else
+  'ng' # unreachable here
+end
+```
+
+### Send a message by moving
+
+`Ractor#send(obj, move: true)` or `Ractor.yield(obj, move: true)` move `obj` to the destination Ractor.
+If the source Ractor touches the moved object (for example, call the method like `obj.foo()`), it will be an error.
+
+```ruby
+# move with Ractor#send
+r = Ractor.new do
+  obj = Ractor.recv
+  obj << ' world'
+end
+
+str = 'hello'
+r.send str, move: true
+modified = r.take #=> 'hello world'
+
+# str is moved, and accessing str from this Ractor is prohibited
+
+begin
+  # Error because it touches moved str.
+  str << ' exception' # raise Ractor::MovedError
+rescue Ractor::MovedError
+  modified #=> 'hello world'
+else
+  raise 'unreachable'
+end
+```
+
+```ruby
+  # move with Ractor.yield
+  r = Ractor.new do
+    obj = 'hello'
+    Ractor.yield obj, move: true
+    obj << 'world'
```

Comment:
We can see that we raise a `Ractor::MovedError` then we rescue `Ractor::RemoteError`. 

With:
```ruby
r = Ractor.new do
  obj = 'hello'
  Ractor.yield obj, move: true
  obj << 'world'
end
str = r.take
begin
  r.take
rescue Ractor::RemoteError
  p str #=> "hello"
end
```
```
ruby 2.8.0dev (2020-07-30T13:57:11Z ractor_parallel 1907c25fc2) [x86_64-darwin18]
#<Thread:0x00007fb12d0661d0 run> terminated with exception (report_on_exception is true):
../../test/ractor_playing.rb:4:in `method_missing': can not send any methods to a moved object (Ractor::MovedError)
	from ../../test/ractor_playing.rb:4:in `block in <main>'
"hello"
```

I am wondering if `Thread.report_on_exception = false` should be something mentioned in this doc, even if it can be a little bit out of context?

-----

Maybe mention that a `Ractor::MovedError` is raised.

```suggestion
    obj << 'world'  # raise Ractor::MovedError
```

At first sight this example can be tricky, especially with the :
```ruby
str = r.take # do not raise
begin
  r.take # raise
```
Maybe adding a little bit of comment could help to better understand what is happening.

Reply:
Maybe the design of `Thread.report_on_exception = false` will be discussed later (not determined now, on default).